### ISSUE TITLE
versioning and rolling upgrades plan

### ISSUE THREAD

**Issue by xiang90** [coreos/etcd#2778]
### bootstrap same version of etcd

`cluster: {a: 2.1, b: 2.1, c :2.1}`

1. after bootstrap the leader sets the cluster version to 2.0.
2. after the first round of successful version detection, the leader sets the cluster version to 2.1.

### bootstrap different versions of etcd

`cluster: {a: 2.2, b:2.1, c: 2.1}`

There are two cases: 

1. `a:2.2` becomes the first leader
2. `b:2.1` or `c:2.1` becomes the first leader

If `a` becomes the first leader:

1. after bootstrap `a:2.2` sets the cluster version to 2.1
2. after the first round of successful detection, the cluster version remains to be 2.1

If `b:2.1` or `c:2.1` becomes the first leader:

1. after bootstrap `b:2.1` or `c:2.1` sets the cluster version to 2.0. (this implies 2.2 needs to support all 2.0 commands. If we want to deprecate a command, we need to stop using the command and wait for at least 2 version bumps before removing the code handling it)
2. after the first round of successful detection, the leader sets the cluster version to 2.1

### member restart

The restarted member recovers the previous cluster version. If all member dies before successfully setting the first cluster version, then it is the same as bootstrap case.

If member restarted with an incompatible version of binary, it dies.

### rolling upgrades

`cluster: {a: 2.0, b:2.0, c: 2.0}` -> cluster version 2.0

upgrades `a:2.1`

`cluster: {a: 2.1, b:2.0, c: 2.0}` -> cluster version 2.0

upgrades `b:2.1`

`cluster: {a: 2.1, b:2.1, c: 2.0}` -> cluster version 2.0

upgrades `c:2.1` -> cluster will be upgraded

`cluster: {a: 2.1, b:2.1, c: 2.1}` -> cluster version 2.1

### new member dynamically joining

there are three cases.

1. the whole cluster is healthy in the view of the member when it joins
    check the versions and cluster versions of all other members, dies if found an incompatibility

    This is 100% safe.

2. some of the members is unhealthy in the view of the member when it joins
    check the versions and cluster versions of all other member, dies if found an incompatibility.

    This is not 100% safe, but 99%.
    
    And the newly joined member might dies at runtime due to runtime version enforcement, though the possibility is low.

    For example: 
    `cluster {a: 2.1, b:2.1, c: 2.2}`
    
    The cluster lost connection with c from the beginning. So the cluster version remains to be 2.0, which is the min cluster version of `a` and `b`.
    
    When `d:2.0` joins, it checks the version of `a:2.1` ,`b:2.1` and the checking result is compatibility. 
    
    However when c comes back, the cluster is in a bad mixed version situation. 
    `d:2.0` will dies with version too low when it checks the version of `c:2.2` successfully. 

3. none of the members is healthy

    do not start

**Comment by xiang90** [comment]
@philips @bakins @yichengq @barakmich 

I enumerate the common cases I can think of. Please take a look.

**Comment by bakins** [comment]
On initial read, LGTM.

**Comment by xiang90** [comment]
```
19:11:32 etcd2 | 2015/04/30 19:11:32 etcdsever: setting up the initial cluster version to 2.0.0
19:11:32 etcd2 | 2015/04/30 19:11:32 etcdsever: set the initial cluster version to 2.0.0
19:11:32 etcd1 | 2015/04/30 19:11:32 etcdsever: set the initial cluster version to 2.0.0
19:11:32 etcd3 | 2015/04/30 19:11:32 etcdsever: set the initial cluster version to 2.0.0
.... rolling upgrade all members...
19:14:39 etcd2 | 2015/04/30 19:14:39 etcdsever: updating the cluster version from 2.0.0 to 2.1.0
19:14:39 etcd2 | 2015/04/30 19:14:39 etcdsever: updated the cluster version from 2.0.0 to 2.1.0
19:14:40 etcd1 | 2015/04/30 19:14:39 etcdsever: updated the cluster version from 2.0.0 to 2.1.0
19:14:40 etcd3 | 2015/04/30 19:14:40 etcdsever: updated the cluster version from 2.0.0 to 2.1.0
```

This is working... So if everyone agrees on this, I get a Pull Request ready for review soon.

**Comment by barakmich** [comment]
That seems good. I'm a little confused why after bootstrapping the cluster is initally 2.0 (is that the case for anything in the 2.x series?)  

Maybe also some thoughts on 2.x -> 3.x as well? Just thinking through the cases.

**Comment by yichengq** [comment]
The lgtm.

One question: what may happen if there exists incompatible version when bootstrap?

**Comment by xiang90** [comment]
@yichengq It will be detected at runtime as dynamic adding case 2. But I guess this should never happen in reality...

**Comment by xiang90** [comment]
closed via #2804
